### PR TITLE
Added `max_bytes` option and FetchRequest_v3 usage.

### PR DESCRIPTION
**Pull request by Drizzt1991**
Also added a check for api_version=(0, 10, 1) and MetadataRequest_v2
[KIP-74](https://cwiki.apache.org/confluence/display/KAFKA/KIP-74%3A+Add+Fetch+Response+Size+Limit+in+Bytes)
NOTE: Depends on PR #974

### CODE REVIEW

**Comment by Drizzt1991** [comment]
Took me long enough =) Fixes #870

File: kafka/consumer/fetcher.py [hunk 1]
```diff
@@ -664,19 +665,34 @@ def _create_fetch_requests(self):
...
+                    -1,  # replica_id
+                    self.config['fetch_max_wait_ms'],
+                    self.config['fetch_min_bytes'],
+                    partition_data.items())
+            else:
+                # As of version == 3 partitions will be returned in order as
+                # they are requested, so to avoid starvation with
+                # `fetch_max_bytes` option we need this shuffle
+                partition_data = list(partition_data.items())
+                random.shuffle(partition_data)
```
Comment (Drizzt1991) [inline kafka/consumer/fetcher.py]: A note to this shuffle. As python dicts and sets are not ordered, currently it will work properly without this shuffle (it will basically be shuffled by _create_fetch_requests, as it uses sets and dicts), but I still prefere it here, as:

* Python3.6 has ordered dicts, which can make the distribution way less equal
* While dicts are not retain insert order, they preserve key order by hashes, which can (probably) result in unequal consumption.

Please correct me if I'm wrong here. My tests run OK even without this shuffle, so I'm a bit concerned.

File: kafka/consumer/group.py [hunk 2]
```diff
@@ -63,6 +63,13 @@ class KafkaConsumer(six.Iterator):
             the server will block before answering the fetch request if
             there isn't sufficient data to immediately satisfy the
             requirement given by fetch_min_bytes. Default: 500.
+        fetch_max_bytes (int): The maximum amount of data the server should
+            return for a fetch request. This is not an absolute maximum, if the
+            first message in the first non-empty partition of the fetch is
+            larger than this value, the message will still be returned to
+            ensure that the consumer can make progress. NOTE: consumer performs
+            multiple fetches in parallel so memory usage will be higher.
```
Comment (jeffwidman) [inline kafka/consumer/group.py]: This note is a great idea, but tweak the wording slightly? It's unclear 1) what it will be higher than , and 2) how much higher... are we talking 20%, 2x, 5x, etc?

File: kafka/protocol/metadata.py [hunk 3]
```diff
@@ -47,6 +47,30 @@ class MetadataResponse_v1(Struct):
     )
 
 
+class MetadataResponse_v2(Struct):
```
Comment (jeffwidman) [inline kafka/protocol/metadata.py]: It's great that you added this, but maybe it should be in a dedicated commit? It just feels logically separate, and I'm unclear why it's needed for the `Fetch_Request/Response` stuff. Personally prefer a commit that says "Here's this new struct, here's what it changes compared to the old one" just because often spelunking in code I'll lookup the commit history to understand changes like this.

File: kafka/consumer/fetcher.py (hunk 1 above)
Comment (jeffwidman) [inline kafka/consumer/fetcher.py]: Good call on 3.6 ordered dicts--makes sense. 

However, if v3 supports returning partitions in the order requested, would it be a good idea to expose this as an optional "feature"? Obviously default to random, but give the option to fetch partitions in the same order they are requested. 

For example if a user eventually wanted to request partitions in a particular order, knowingly risking starvation, but wanting to use it as a poor-man's prioritization queue? 

Just tossing it out there... it might be a bad idea, and also might require a lot of internal changes in `kafka-python`, I haven't spent enough time thinking about it.

File: kafka/consumer/fetcher.py [hunk 4]
```diff
@@ -664,19 +665,34 @@ def _create_fetch_requests(self):
...
+            if version < 3:
+                requests[node_id] = FetchRequest[version](
+                    -1,  # replica_id
+                    self.config['fetch_max_wait_ms'],
+                    self.config['fetch_min_bytes'],
+                    partition_data.items())
+            else:
+                # As of version == 3 partitions will be returned in order as
+                # they are requested, so to avoid starvation with
+                # `fetch_max_bytes` option we need this shuffle
```
Comment (jeffwidman) [inline kafka/consumer/fetcher.py]: Maybe also add to the comment something about Python 3.6 consistent ordering of dicts...

> ...this shuffle. Otherwise, in Python >=  3.6 dicts have consistent order, so can't guarantee hashing to dict/set will randomize order of partitions.

File: kafka/consumer/fetcher.py (hunk 1 above)
Comment (jeffwidman) [inline kafka/consumer/fetcher.py]: Weird, Github swallowed my earlier comment. 

+1 for explicitly shuffling to avoid py3.6 issues. 

However, since the protocol explicitly responds in the order requested, maybe this should be exposed to the user? Default to shuffling the order to avoid starvation, but allow the user to override that to always request a particular order... I can see someone using this as a sort of poor-man's priority-queue.  Not sure if this is a good idea, or if it'd require reworking a lot of of `kafka-python` internals... just a thought.

Comment (Drizzt1991) [inline kafka/consumer/fetcher.py]: I did have the same idea, but could not find any applications to this. And it's probably not this PR's responsibility, as it's rather a feature.

File: kafka/protocol/metadata.py (hunk 3 above)
Comment (jeffwidman) [inline kafka/protocol/metadata.py]: Broken out in #974.

File: kafka/consumer/group.py [hunk 5]
```diff
@@ -63,6 +63,14 @@ class KafkaConsumer(six.Iterator):
             the server will block before answering the fetch request if
             there isn't sufficient data to immediately satisfy the
             requirement given by fetch_min_bytes. Default: 500.
+        fetch_max_bytes (int): The maximum amount of data the server should
+            return for a fetch request. This is not an absolute maximum, if the
+            first message in the first non-empty partition of the fetch is
+            larger than this value, the message will still be returned to
+            ensure that the consumer can make progress. NOTE: consumer performs
+            fetches to multiple nodes in parallel so memory usage will depend
+            on the number of nodes containing partitions for the topic.
```
Comment (jeffwidman) [inline kafka/consumer/group.py]: Can we call them 'brokers' rather than 'nodes'?

**Comment by jeffwidman** [comment]
LGTM with the caveat that I didn't have time to try the code, just read through it.

File: kafka/conn.py [hunk 6]
```diff
@@ -778,7 +778,8 @@ def filter(self, record):
         log.addFilter(log_filter)
 
         test_cases = [
-            ((0, 10), ApiVersionRequest[0]()),
+            ((0, 10, 1), MetadataRequest[2]([])),
```
Comment (dpkp) [inline kafka/conn.py]: note to self: for any broker after 0.10 we should just use the results of ApiVersions api request to set api version support.

File: kafka/protocol/metadata.py [hunk 7]
```diff
@@ -47,6 +47,30 @@ class MetadataResponse_v1(Struct):
...
+    API_KEY = 3
+    API_VERSION = 1
```
Comment (dpkp) [inline kafka/protocol/metadata.py]: should be API_VERSION = 2 ?

File: kafka/protocol/metadata.py (hunk 3 above)
Comment (dpkp) [inline kafka/protocol/metadata.py]: agree -- would prefer to separate this into separate PR.

Comment (Drizzt1991) [inline kafka/protocol/metadata.py]: It's separated, see #974

File: kafka/protocol/metadata.py (hunk 7 above)
Comment (Drizzt1991) [inline kafka/protocol/metadata.py]: Ou, yea =)

**Comment by Drizzt1991** [comment]
As for the failing tests, I do suspect that something is strange, as they fail quite strangely on py27 and py26 on KAFKA<0.10.1, but the fail message has no sense:
```
ERROR: InvocationError: '/home/travis/build/dpkp/kafka-python/.tox/py27/bin/py.test --pylint --pylint-rcfile=pylint.rc --pylint-error-types=EF --cov=kafka --cov-config=.covrc'
```

And the tests run OK on my machine, I have no clue.

File: kafka/conn.py (hunk 6 above)
Comment (Drizzt1991) [inline kafka/conn.py]: Should I implement it?

Comment (dpkp) [inline kafka/conn.py]: ah yes. forgot about this little wart of kafka brokers. so it turns out that kafka < 0.10 will silently ignore incorrect metadata api version requests (or at least not close the socket as expected). The socket behavior is different between py2 and py3, causing KafkaConnection._recv() to block indefinitely on py2.

So we can't use MetadataRequest[2] to check version and will need to implement the smarter ApiVersion approach to avoid breaking auto version checks on older brokers.

Comment (dpkp) [inline kafka/conn.py]: test via `KAFKA_VERSION=0.8.0 tox -e py27 -- test/test_consumer_group.py::test_paused` -- this hangs indefinitely on my laptop

Comment (Drizzt1991) [inline kafka/conn.py]: ok, will try it.

**Comment by Drizzt1991** [comment]
Changed check_version to use ApiVersionResponse result. Works quite good for me.

**Comment by Drizzt1991** [comment]
@dpkp Is there anything else for this PR, that needs to be addressed?

File: kafka/conn.py [hunk 8]
```diff
@@ -830,6 +849,10 @@ def connect():
                 self._sock.setblocking(False)
 
             if f.succeeded():
+                if version == (0, 10):
```
Comment (dpkp) [inline kafka/conn.py]: better would be `if isinstance(request, ApiVersionRequest[0]):`

File: kafka/conn.py [hunk 9]
```diff
@@ -752,6 +753,24 @@ def _next_correlation_id(self):
         self._correlation_id = (self._correlation_id + 1) % 2**31
         return self._correlation_id
 
+    def _check_version_above_0_10(self, response):
+        test_cases = [
+            # format (<broker verion>, <needed struct>)
```
Comment (dpkp) [inline kafka/conn.py]: might note that the order here matters, and/or make sure we reverse sort it before checking for best match below

**Comment by dpkp** [comment]
Looks great. I added two very minor points that we can fix later (or you can address now if you have time). I'll merge this before next release.

**Comment by Drizzt1991** [comment]
Thanks for the cleanup, looks good =) I missed it on the weekend
